src/System.Memory/src/System/Buffers/ReadOnlyBuffer_helpers.cs

```diff
@@ -0,0 +1,305 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.CompilerServices;
+
+namespace System.Buffers
+{
+    public readonly partial struct ReadOnlyBuffer<T>
+    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal bool TryGetBuffer(SequencePosition start, SequencePosition end, out ReadOnlyMemory<T> data, out SequencePosition next)
+        {
+            if (start.Segment == null)
+            {
+                data = default;
+                next = default;
+                return false;
+            }
+
+            var startIndex = start.Index;
+            var endIndex = end.Index;
+            var type = GetBufferType();
+
+            startIndex = GetIndex(startIndex);
+            endIndex = GetIndex(endIndex);
+
+            switch (type)
+            {
+                case BufferType.MemoryList:
+                    var bufferSegment = (IMemoryList<T>) start.Segment;
+                    var currentEndIndex = bufferSegment.Memory.Length;
+
+                    if (bufferSegment == end.Segment)
+                    {
+                        currentEndIndex = endIndex;
+                        next = default;
+                    }
+                    else
+                    {
+                        var nextSegment = bufferSegment.Next;
+                        if (nextSegment == null)
+                        {
+                            if (end.Segment != null)
+                            {
+                                ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();
+                            }
+
+                            next = default;
+                        }
+                        else
+                        {
+                            next = new SequencePosition(nextSegment, 0);
+                        }
+                    }
+
+                    data = bufferSegment.Memory.Slice(startIndex, currentEndIndex - startIndex);
+
+                    return true;
+
+
+                case BufferType.OwnedMemory:
+                    var ownedMemory = (OwnedMemory<T>) start.Segment;
+                    data = ownedMemory.Memory.Slice(startIndex, endIndex - startIndex);
+
+                    if (ownedMemory != end.Segment)
+                    {
+                        ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();
+                    }
+
+                    next = default;
+                    return true;
+
+                case BufferType.Array:
+                    var array = (T[]) start.Segment;
+                    data = new Memory<T>(array, startIndex, endIndex - startIndex);
+
+                    if (array != end.Segment)
```

Comment:
I don't know how to reason about this check. It looks like it expects start.Segment == end.Segment always. Otherwise, we throw. When do we catch these exceptions?